# Защита персональных данных клиентов / Protection of clients' personal data

Требуется защитить данные клиентов страховой компании, разработать такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось.

It is necessary to protect the data of the insurance company's clients, to develop a method of data transformation so that it is difficult to recover personal information from it.

It is necessary to protect the data so that the quality of machine learning models does not deteriorate during conversion.

## Загрузка данных / Loading Data

In [1]:
import pandas as pd
import numpy as np
from sklearn.metrics import r2_score

In [2]:
data = pd.read_csv('/Users/vladamalkina/Downloads/insurance-2.csv')
display(data)
data.info()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0       1     41.0   49600.0            1                  0
1       0     46.0   38000.0            1                  1
2       0     29.0   21000.0            0                  0
3       0     21.0   41700.0            2                  0
4       1     28.0   26100.0            0                  0
...   ...      ...       ...          ...                ...
4995    0     28.0   35700.0            2                  0
4996    0     34.0   52400.0            1                  0
4997    0     20.0   33900.0            2                  0
4998    1     22.0   32700.0            3                  0
4999    1     28.0   40600.0            1                  0

[5000 rows x 5 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


In [3]:
features = data.drop('Страховые выплаты', axis=1)
target = data['Страховые выплаты']

В данных отсутствуют пропуски, следовательно можно переходить к этапу машинного обучения.

Страховые выплаты - целевой признак.

Пол, Возраст, Зарплата, Члены семьи - признаки.

There are no gaps in the data, so you can move on to the machine learning stage.

Insurance payments are the target attribute.

Gender, Age, Salary, Family members - signs.

## Алгоритм преобразования / Conversion algorithm

**Алгоритм / Algorithm**

1) Обучение линейной регрессии / Linear Regression Training;

2) Создание обратимой матрицы / Creating an Invertible Matrix;

3) Умножение признаков на эту матрицу / Multiplying features by this matrix;

4) Вычисление r2 для новой модели / Calculating r2 for the new model.

**Обоснование / Rationale**

При умножении признаков на обратимую матрицу, сами признаки не меняются; мы переходим в новую систему. Поэтому качество модели не изменится.

When multiplying features by an invertible matrix, the features themselves do not change; We are moving to a new system. Therefore, the quality of the model will not change.

## Проверка алгоритма / Algorithm verification

Проверка модели без умножения на обратимую матрицу:

Model checking without multiplication by an invertible matrix:

In [4]:
class LinearRegression:
    def fit(self, train_features, train_target):
        X = np.concatenate((np.ones((train_features.shape[0], 1)), train_features), axis=1)
        y = train_target
        w = np.linalg.inv(X.T@X)@X.T@y
        self.w = w[1:]
        self.w0 = w[0]

    def predict(self, test_features):
        return test_features.dot(self.w) + self.w0
    
    
model = LinearRegression()
model.fit(features, target)
predictions = model.predict(features)
print(r2_score(target, predictions))

0.42494550286668


Проверка модели с умножением на обратимую матрицу:

Checking the model with multiplication by an invertible matrix:

In [5]:
model = LinearRegression()
matrix = np.random.normal(loc = 0.0, scale = 2.0, size = (4, 4))
np.linalg.inv(matrix)
features = features @ matrix
model.fit(features, target)
predictions = model.predict(features)
print(r2_score(target, predictions))

0.424945502866535


Качество модели до и после обучения не изменилось (r2 = 0.425)

The quality of the model before and after training did not change (r2 = 0.425)